rpcclient/client.go

```diff
@@ -0,0 +1,751 @@
+// Package rpcclient is used to do jsonrpc calls with Flashbots request signatures (X-Flashbots-Signature header)
+//
+// Implemenation and  interface is a slightly modified copy of  https://github.com/ybbus/jsonrpc
+// The differences are:
+// * we handles a case when Flashbots API returns errors incorrectly according to jsonrpc protocol (backwards compatibility)
+// * we can sign request body with ecdsa
+package rpcclient
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"reflect"
+	"strconv"
+
+	"github.com/flashbots/go-utils/signature"
+)
+
+const (
+	jsonrpcVersion = "2.0"
+	// This error code is set in the error reponse when interacting with Flashbots RPC that has broken error response
+	// see docs for RPCClientOpts.RejectBrokenFlashbotsErrors
+	FlashbotsBrokenErrorResponseCode = -32088
+)
+
+// RPCClient sends JSON-RPC requests over HTTP to the provided JSON-RPC backend.
+//
+// RPCClient is created using the factory function NewClient().
+type RPCClient interface {
+	// Call is used to send a JSON-RPC request to the server endpoint.
+	//
+	// The spec states, that params can only be an array or an object, no primitive values.
+	// So there are a few simple rules to notice:
+	//
+	// 1. no params: params field is omitted. e.g. Call(ctx, "getinfo")
+	//
+	// 2. single params primitive value: value is wrapped in array. e.g. Call(ctx, "getByID", 1423)
+	//
+	// 3. single params value array or object: value is unchanged. e.g. Call(ctx, "storePerson", &Person{Name: "Alex"})
+	//
+	// 4. multiple params values: always wrapped in array. e.g. Call(ctx, "setDetails", "Alex, 35, "Germany", true)
+	//
+	// Examples:
+	//   Call(ctx, "getinfo") -> {"method": "getinfo"}
+	//   Call(ctx, "getPersonId", 123) -> {"method": "getPersonId", "params": [123]}
+	//   Call(ctx, "setName", "Alex") -> {"method": "setName", "params": ["Alex"]}
+	//   Call(ctx, "setMale", true) -> {"method": "setMale", "params": [true]}
+	//   Call(ctx, "setNumbers", []int{1, 2, 3}) -> {"method": "setNumbers", "params": [1, 2, 3]}
+	//   Call(ctx, "setNumbers", 1, 2, 3) -> {"method": "setNumbers", "params": [1, 2, 3]}
+	//   Call(ctx, "savePerson", &Person{Name: "Alex", Age: 35}) -> {"method": "savePerson", "params": {"name": "Alex", "age": 35}}
```

Comment:
Hmm, I thought that Ethereum's "flavor" of JSONRPC always expected `params` to be an array?